### PR TITLE
Team: Randy Dang, Perry Tran, Dustin Dubia

### PR DESCRIPTION
**Pull request by dangrt5**

### CODE REVIEW

**Comment by TimelikeClosure** [comment]
Nice job overall with specific commit messages (minus one offending `fixed bugs`). In general though, I would recommend using normal English syntax over camelCase, and describing what the changes accomplish conceptually (ex. "Prevented players from selecting previously-filled squares") instead of writing the specific code changes (ex. "Added 'notAvailable' class").

File: index.html [hunk 1]
```diff
@@ -0,0 +1,53 @@
+<!DOCTYPE html>
+<html lang="en" dir="ltr">
+  <head>
+    <meta charset="utf-8">
+    <title>Tic Tac Toe - An Interactive and Competitive Game</title>
+    <script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
+    <link rel="stylesheet" href="style.css">
+    <script src="main.js"></script>
+  </head>
```
Comment (TimelikeClosure) [inline index.html]: To make implementing mobile responsiveness easier, make sure you remember to add the `<meta name="viewport" ... >` tag to the `<head>`. It will save you many `<head>`aches later.

File: index.html [hunk 2]
```diff
@@ -0,0 +1,53 @@
+<!DOCTYPE html>
+<html lang="en" dir="ltr">
+  <head>
+    <meta charset="utf-8">
```
Comment (TimelikeClosure) [inline index.html]: Nice job remembering to specify the document's language and character set.

File: index.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  </head>
+  <body>
+    <div class="container">
+
+      <div class="header">
+        <h1>Tic Hack Toe</h1>
+      </div>
+
+      <div class="content">
+      <div class="gameBoard"></div>
```
Comment (TimelikeClosure) [inline index.html]: Make sure you remember to consistently indent child elements.

File: index.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      </div>
+
+        <div class="resetbuttonContainer">
+          <div class="resetButtons">
+
+            <button gridSize="3">3x3 Squares</button>
+            <button gridSize="4">4x4 Squares</button>
+            <button gridSize="5">5x5 Squares</button>
```
Comment (TimelikeClosure) [inline index.html]: With custom attributes, you'll want to preface them with `data-` for SEO and Accessibility reasons. For example, these attributes would be named `data-gridSize`.

File: index.html [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          <h1>
+            [<span class='playerOneTurn'>X</span>]
+            <span>Player 1</span>:
+            <span id="player1Score">0</span>
+          </h1>
+          <h1>
+            [<span class='playerTwoTurn'>O</span>]
+            <span>Player 2</span>:
+             <span id="player2Score">0</span>
+          </h1>
```
Comment (TimelikeClosure) [inline index.html]: Rather than using a whole set of specific classes & ids per player, try adding an id each parent, then a shared set of classes to the children. For example, the above could be written as follows:
```
<h1 id="player1Status">
  [<span class='playerTurn'>X</span>]
 <span>Player 1</span>:
  <span class="playerScore">0</span>
</h1>
<h1 id="player2Status">
  [<span class='playerTurn'>O</span>]
 <span>Player 1</span>:
  <span class="playerScore">0</span>
</h1>
```
Notice the child classes are shared between both players, but you can still target them uniquely by referencing the specific parent if need be (ex. `#player1Status>.playerScore`).

File: index.html [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+      </div>
+      <div id="myModal" class="modalShadow">
```
Comment (TimelikeClosure) [inline index.html]: Never ever _ever_ name something `my`-anything! Not only is it horrible naming practice, but it also screams "I copy this code from somewhere without trying to understand it".

File: index.html [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        <div class="modal">
+          <p id="winnerInformation">Winner information.</p>
+        </div>
+      </div>
+    </div>
+
+    <audio class="soundOnLoad" src="sounds/soundOnLoad.mp3" autoplay preload></audio>
```
Comment (TimelikeClosure) [inline index.html]: Do you know what `preload` does? Are you able to explain what it does to someone else? Why are you using it here?

File: style.css [hunk 8]
```diff
@@ -0,0 +1,140 @@
+@font-face {
+  font-family: alphaFont;
+  src: url("fonts/font2.otf");
```
Comment (TimelikeClosure) [inline style.css]: Nice use of `@font-face` for adding a custom font. Your file naming, however, needs improvement...

File: style.css [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    margin: 0;
+    font-size: 0;
+    background: url("images/background.jpg");
+    background-size: 110% auto;
+    background-repeat: no-repeat;
+}
+
+.header {
+    text-align: center;
+    font-size: 2vw;
```
Comment (TimelikeClosure) [inline style.css]: Be careful of sizing fonts in viewport units. They have a habit of either making text unreadably-small on small screens or far-too-big on extra large screens. Generally, you want to use either fixed pixels (`px`) or dynamic sizing not directly based on screen size (`em`, `rem`, `%`, etc).

File: style.css [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+.header h1 {
+  margin: 0;
+  font-family: alphaFont;
+}
+
+.content {
+    width: 50%;
+    height: 70vh;
```
Comment (TimelikeClosure) [inline style.css]: When you need content to fill the page, but not shrink too much when the viewport height is small (for example, on landscape phones), use `min-height` instead of `height`.

File: style.css [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+.row .column:first-of-type {
+  border-left: 0;
+}
+
+div.row {
+    top: 0;
+    left: 0;
+    width: 100%;
```
Comment (TimelikeClosure) [inline style.css]: Your rows are already `display: block`, which normally takes up the full width of their parents. Setting the `width` to `100%` here does nothing.

File: style.css [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    display: inline-block;
+    /* background-color: lawngreen; */
+    border: 1px solid black;
+    font-size: 8vw;
+    text-align: center;
+    vertical-align: top;
+}
+
+.resetbuttonContainer {
+  /* border: solid black 1px; */
```
Comment (TimelikeClosure) [inline style.css]: Don't forget to remove commented code when you're done with it.

File: style.css [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+.resetButtons {
+
+    text-align: center;
+}
+
+.resetButtons button {
+    font-family: alphaFont;
+    src: url("fonts/font2.otf");
```
Comment (TimelikeClosure) [inline style.css]: You already include the font `src` in the `@font-face` rule up above. Why include it here as well?

File: style.css [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+.scoreAndPlayers h1 {
+  display: inline-block;
+  margin: 1% 3%;
+}
+
+.scoreAndPlayers span {
+  font-family: alphaFont;
+}
+
+.notAvailable {}
```
Comment (TimelikeClosure) [inline style.css]: Styling for `.notAvailable` not available.

File: style.css [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    display: none;
+    position: fixed;
+    z-index: 1;
+    left: 0;
+    top: 0;
+    width: 100%;
+    height: 100%;
+    overflow: auto;
+    background-color: rgb(0,0,0);
+    background-color: rgba(0,0,0,0.4); /* Black w/ opacity */
```
Comment (TimelikeClosure) [inline style.css]: Setting the same property twice causes the first to be overwritten.

File: main.js [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+function populateGameBoardArray(boardSize) {
+    // resets gameboard Array, no visible front-end effect
+    gameBoardSize = boardSize;
+    gameBoardArray = [];
+    var counter = 10;
+    for (var i = 0; i < boardSize; i++) {
+        gameBoardArray.push([]);
+        for (var x = 0; x < boardSize; x++) {
+            gameBoardArray[i].push(counter);
+            counter++;
```
Comment (TimelikeClosure) [inline main.js]: What's the purpose of populating the game board state with gradually-increasing numbers?

File: main.js [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+}
+
+function createGameBoard(boardSize) {
+    currentPlayer = 1;
+    populateGameBoardArray(boardSize);
+    winCondition = boardSize;
```
Comment (TimelikeClosure) [inline main.js]: As an aside, it becomes much more difficult to win on larger game boards in Tic-Tac-Toe if the win condition requires reaching all the way across it. On the other side, it becomes much easier on larger boards if the win condition is always kept the same (3-in-a-row). Usually, the balance point is somewhere in-between (which explains, for example, why Connect 4 is done on a 6 x 7 grid).

File: main.js [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+    else if (boardSize === "5") {
+      $(".gameBoard").css({
+        "width": "52vw",
+        "height": "41vh"
+      });
+      $(".square").css({
+        "width": "19.1%",
+        "height": "100%",
+        "font-size": "5vw"
+      });
```
Comment (TimelikeClosure) [inline main.js]: Manually setting CSS inside the script prevents non-JS-speaking designers from easily modifying style, and is harder to find, since it's not present on the stylesheet with the rest of the styling. A better way of handling it would be setting a class or other attribute on the ancestor (`.gameBoard`) and styling the board & squares based upon which one is used.

For example, suppose we set a `data-size` property on `.gameBoard` using `$('.gameBoard').attr('data-size', boardSize);`. Then we could style it within the CSS stylesheet like:
```
.square {
  height: 100%; // always the same
}
.gameBoard[data-size=3] {
  width: 40vw;
  height: 65vh;
}
.gameBoard[data-size=3] .square {
  width: 33.33%;
  font-size: 9vw;
}
```

File: main.js [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+function initializeApp() {
+    addEventListeners();
+}
+
+function addEventListeners() {
+    $('.gameBoard').on('click', '.square', playerOneAndTwo);
+    $(".resetButtons").click(clickSound);
+    $('.resetButtons:not(.reset)').on('click', 'button' ,gridSize);
```
Comment (TimelikeClosure) [inline main.js]: Your click handlers probably do not run when you think they do...

screenshot_24

More specifically, the `gridSize` function still runs when the `<button class="reset">` is clicked. The reason for this is `$('.resetButtons:not(.reset)')` selects all elements with the class `resetButtons` that do not themselves have the class `reset`. Since `<div class="resetButtons">` has the class `resetButtons` and no `reset` class, the event handler is applied to it. Since it is a delegated event handler listening for clicks on descendents matching the `button` selector, it runs on all buttons, _including the `.reset` button_.

TL:DR - Your `gridSize` function runs when clicking on the `.reset` button and here's why:

File: main.js [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+        return;
+    }
+
+    $(this).addClass("notAvailable");
+    if (currentPlayer === 1) {
+        currentSquareClicked.text('X');
+    } else {
+        currentSquareClicked.addClass("O");
+        currentSquareClicked.text('O');
+    }
```
Comment (TimelikeClosure) [inline main.js]: Decided not to make use of the beautiful green provided by the `X` class?

File: main.js [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+            var spanId = $("#player2Score");
+            spanId.text(parseFloat(spanId.text())+1);
+        }
+    } else {
+        $("#winnerInformation").text("No Winner!");
+    }
+    modalDiv.css("display","block");
+    modalDiv.click( function () {
+        modalDiv.css("display","none");
+    })
```
Comment (TimelikeClosure) [inline main.js]: Ack, more manual CSS application!

File: main.js [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+    }
+    if (occupiedSquares===gameBoardSize*gameBoardSize) {
+        reportGameEnded(-1);
+    }
+}
+
+function reportGameEnded (currentPlayer) {
+    var modalDiv = $("#myModal");
+    if (currentPlayer>-1) {
+        $("#winnerInformation").text("Winner is " + displayPlayerNumber());
```
Comment (TimelikeClosure) [inline main.js]: Based on the name, I would expect `displayPlayerNumber()` to modify the DOM itself.

File: main.js [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+  playSound(sounds.click);
+}
+
+function currentPlayersTurn() {
+    if (currentPlayer === 1) {
+        $('.playerOneTurn').css('color', '#955bbc');
+        $('.playerTwoTurn').css('color', 'black');
+    } else {
+        $('.playerOneTurn').css('color', 'black');
+        $('.playerTwoTurn').css('color', '#955bbc');
```
Comment (TimelikeClosure) [inline main.js]: And even more manual CSS changes. Also, switching back and forth between named colors and hex codes is giving me whiplash.

File: main.js [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+        checkLimits = gameBoardSize - nColumns -1;
+    }
+    for (var i = 1; i<= checkLimits; i++) {
+        if (gameBoardArray[nColumns+i][nRows-i]=== playerNumber) {
+            currentVectorSequence ++;
+            if (currentVectorSequence > highestVectorSequence) {highestVectorSequence=currentVectorSequence}
+        } else {
+            break;
+        }
+    }
```
Comment (TimelikeClosure) [inline main.js]: Your win checking seems to work, but is a very brute-force way: checking each direction manually means you've written nearly the same code eight times. This could be greatly reduced by:
1. Combining opposite direction checks into a single check. (e.g. Finding the longest sequence from the extreme left to the extreme right).
2. Combining each of the four directions (vertical, horizontal, slant-left, slant-right) into a single function that takes in a starting position and direction (ex. `[nColumns, 0]` for position, `[0, 1]` for direction).

File: main.js [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+    });
+}
+
+function playerOneAndTwo() {
+    var currentSquareClicked = $(this);
+    var columnCoordinate = parseInt(currentSquareClicked.attr('columns'));
+    var rowCoordinate = parseInt(currentSquareClicked.parent().attr('rows'));
+
+
+    if ($(this).hasClass("notAvailable")) {
```
Comment (TimelikeClosure) [inline main.js]: This could probably run before the coordinates are retrieved.
